### ISSUE TITLE
Drag & Drop Re-ordering needs to account for submenus

### ISSUE THREAD

**Issue by celloexpressions** [voldemortensen/menu-customizer#4]
Keyboard-accessible reordering needs to account for moving parent items as well, to move the submenu items with it.

**Comment by celloexpressions** [comment]
This desperately needs someone to dig in to it.

**Comment by celloexpressions** [comment]
To put it another way, without this there is absolutely no way that we can propose Menu Customizer for core. This is probably the biggest single task we have left, and pretty much the only remaining thing for feature parity with the existing menus screen. Anyone up to take a stab at it? This is the reason for the large section of commented-out code in menu-customizer.js, use nav-menus.js in core as inspiration.

**Comment by voldemortensen** [comment]
I've been trying to get this one worked out. No luck so far.

**Comment by designsimply** [comment]
This problem showed up in a user test. Here's the related video snippet for reference: [1m48s](https://cloudup.com/cQAyyZtg1C9) 

Aside: it's interesting that he opened the item and looked for an option inside it to make it a child of another item.

<sup>Pulled from the test at https://make.wordpress.org/design/2015/05/27/customizer-menus-user-testing/#comment-24592 using WP 4.3-alpha-32600 and the Menu Customizer 0.4 (6a6fca3).</sup>

**Comment by designsimply** [comment]
Note: In past user testing for menus, people said they really liked that there are dashed outlines for dropzone target areas that take up space so they're easy to see as you're moving items around on the existing menus admin page: [29s screencast](https://cloudup.com/cGRxVt0JYkK)

Can dashed dropzone target areas be added to customizer menus for feature parity?

**Comment by voldemortensen** [comment]
@designsimply, that's in the feature/sub-menus branch, it just doesn't work yet :) We're definitely adding that.

**Comment by designsimply** [comment]
@voldemortensen, brilliant, thanks! It's a really nice part of the UX.

**Comment by westonruter** [comment]
As noted in Slack: https://wordpress.slack.com/archives/core-customize/p1433203971002514

> There are glitches with submenus which I haven't been able to figure out yet. If there is anyone who is an authority on wpNavMenu and how that all works with jQuery UI Sortables, it would be very helpful if they could look into that. I think my time would be better spent now working on 67 and some of the underlying Customizer architectural issues, rather than frontend JS/CSS problems. I should probably pass it off to someone else.

So I am unassigning this from myself.

**Comment by adamsilverstein** [comment]
@westonruter I'll try to help here so you can focus on 67, which is a bit deep for my first customizer internals foray :) 

Pulled the branch and see the submenus working; here are the issues I see so far, please add any other issues you know about here; these may be existing issues?

* Pretty easy to drag an item out of menu and remove it inadvertantly
* Was able to drop submenu at top making it a child of nothing: http://cl.ly/image/3Y231T3m221U
* Was able to get grandchildren below top level without children with some rearranging: http://cl.ly/image/1y301d0J2m2k

**Comment by westonruter** [comment]
Thanks very much @adamsilverstein.

Another issue is that I could drag a menu item below the whole menu entirely, dropping a menu item among the checkboxes below :exclamation: 

Perhaps some of these issues are caused by our seeming inability to override the sortable options:

```js
	menuList.sortable( 'option', {
		items: '> .customize-control-menu_item',
		connectWith: '.accordion-section-content:has(.customize-control-menu_item)'
	} );
	// @todo these option overrides don't seem to be applying.
```

**Comment by adamsilverstein** [comment]
Yea, I just saw that issue as well, I'm working on it :) Thanks for the
@todo tip.

On Tue, Jun 2, 2015 at 11:00 AM, Weston Ruter <notifications@github.com>
wrote:

> Thanks very much @adamsilverstein <https://github.com/adamsilverstein>.
>
> Another issue is that I could drag a menu item below the whole menu
> entirely, dropping a menu item among the checkboxes below [image:
> :exclamation:]
>
> Perhaps some of these issues are caused by our seeming inability to
> override the sortable options:
>
>     menuList.sortable( 'option', {
>         items: '> .customize-control-menu_item',
>         connectWith: '.accordion-section-content:has(.customize-control-menu_item)'
>     } );
>     // @todo these option overrides don't seem to be applying.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/voldemortensen/menu-customizer/issues/4#issuecomment-108015094>
> .
>



-- 
Adam Silverstein
Engineering Manager @ 10up

**Comment by designsimply** [comment]
When you drag and drop and item into a new location, can you stop the item from expanding after it's dropped? It's disruptive in that context.

**Comment by adamsilverstein** [comment]
Ok, I see this in firefox, will fix.

**Comment by adamsilverstein** [comment]
@designsimply The opening when dragging issue in firefox is corrected in my latest push.

**Comment by designsimply** [comment]
@adamsilverstein, cool, I re-tested with 466b0f4 on the feature/sub-menus branch and confirmed that items now stay closed after drag & drop. Thanks! :+1:

**Comment by adamsilverstein** [comment]
I spent some more time digging into this today.

Most of the issues we are seeing are related to having additional li elements inside the container containing the sortable items - the element we apply sortable() to.

I spent a bunch of time trying to track down the bug where you drag menu items and they disappear - this  happens with both the regular items and menus with submenus, especially when dragging to the bottom or top of the list - its pretty easy to reproduce.

To test my idea that the other elements inside the dif interfere, I temporarily removed this section in this commit: https://github.com/voldemortensen/menu-customizer/commit/5b51cdb9826774a4a38062fefcc3817599d384c6

In my testing, removing the extra elements from the bottom of the pane solves all the weird disappearing items glitches. I still have some issues placing items exactly where I want, I think these same issues exist in the current menu sortable implementation. I've also been testing with tolerance: 'pointer'.

**Comment by adamsilverstein** [comment]
I'm still digging here and making some progress. It turns out that we can't easily override the items selected by nav-menu.js in its sirtable setup - in part due to an open bug in jQuery UI sortable: http://bugs.jqueryui.com/ticket/8532. tl;dr the <li> elements below the menu list wind up becoming valid sortable elements and targets, wreaking havoc on the sortable functionality. 

Now looking at slightly extending initSortables to allow for our use case.

**Comment by westonruter** [comment]
@adamsilverstein good find! So perhaps we should let the item selector be defined on `wpNavMenu.options` instead so that it can be easily overridden?


```js
wpNavMenu = {
	options : {
		menuItemDepthPerLevel: 30, // Do not use directly. Use depthToPx and pxToDepth instead.
		globalMaxDepth: 11,
		sortableHandleSelector: '.menu-item-handle',
	},
	/* ... */
};
```

And then later:

```js
api.menuList.sortable({
	handle: api.options.sortableHandleSelector,
	placeholder: 'sortable-placeholder',
	/* ... */
```

Then, for Menu Customizer, we would just have to override `wpNavMenu.options.sortableHandleSelector` up front.

If you like, you can amend Core via this existing PR which I am using for #67: https://github.com/xwp/wordpress-develop/pull/91

**Comment by adamsilverstein** [comment]
Yes, that is the direction i was heading - adding a variable for 'items' we can override up front. Testing that locally now.

**Comment by adamsilverstein** [comment]
This helped get the sortable() items right. Wound up taking a bit more tinkering with nav-menu.js internals - in particular, adding a '.menu-item' target to next() and prev() in updateSharedVars() to prevent the extra `<li>`s from being included in the drag and drop action. 

Working well locally now, doing some more testing then will push a patch/branch for testing.

**Comment by adamsilverstein** [comment]
Ok, this is working much better now! I still some issues and this deserves more testing.

To test the latest version in [this commit](https://github.com/voldemortensen/menu-customizer/commit/2d363729782ddcafb4ea543a15bb47bbfc7e6fc9), you also need to apply this patch to core: https://cloudup.com/cy12FAJWrRY - 

The patch lets us add an items option to wpNavMenu that is then used internally by nav-menus.js for the `items` parameter when setting up the sortable(). In addition, I added some additional targeting in several places to ensure our extra `li` tags aren't included in the logic. This fixes most/all of the dragging issues - still deserves more testing. I also tested the old interface to be sure nothing changed there.
